### PR TITLE
flock: Use fcntl constants directly from libc crate on Unix targets

### LINKED ISSUE #57007: File-locking API uses incorrect fnctl constants on at least sparc64
As described in https://github.com/rust-lang/rust/issues/49773#issuecomment-449006397, the file-locking API uses incorrect fnctl constants on sparc64.

Compare the glibc headers on sparc64:

> https://github.molgen.mpg.de/git-mirror/glibc/blob/master/sysdeps/unix/sysv/linux/sparc/bits/fcntl.h#L53

and what's in the rust sources:

> https://doc.rust-lang.org/nightly/nightly-rustc/src/rustc_data_structures/flock.rs.html#45

Would there be a mechanism to sync the constants directly with the glibc sources, i.e. "using the headers"?

@jrtc27 @karcherm @psumbera @michaelwoerister @alexcrichton

### PR DESCRIPTION
**Pull request by glaubitz**
Since the values for the fcntl constants can vary from architecture
to architecture, it is better to use the values defined in the libc
crate instead of assigning literals in the flock code which would
make the assumption that all architectures use the same values.

Fixes #57007

### CODE REVIEW

**Comment by glaubitz** [comment]
This also requires two small fixes in the libc crate: https://github.com/rust-lang/libc/pull/1186

**Comment by nagisa** [comment]
> This also requires two small fixes in the libc crate: rust-lang/libc#1186

liblibc is a submodule in rust tree, so you will have to update the submodule after the referenced PR lands and before this can/should land.

**Comment by glaubitz** [comment]
Thanks for the heads-up! I will open a PR for that, too.

**Comment by nagisa** [comment]
Would be fine to just have it all be in the same PR (i.e. this PR), unless you’re in a hurry to land this regardless of whether the `liblibc` change langs.

**Comment by varkor** [comment]
> And I'm not in a hurry, I'll prepare the PR whatever suits you best.

I think it would be most straightforward just to update the submodule in this PR, seeing as how they're directly related.

**Comment by glaubitz** [comment]
> I think it would be most straightforward just to update the submodule in this PR, seeing as how they're directly related.

And I'm just trying to find out how to do that properly. I have updated a submodule in a PR before.

Care to explain?

**Comment by nagisa** [comment]
Oh, it seems that we have recently stopped using the submodule approach and are now pulling the dependency in via cargo from crates.io instead... Which essentially means that all you need to do is to update the relevant `Cargo.toml` files to use the new version from crates.io and it should work just as one would expect.

cc @alexcrichton liblibc probably needs a new release, right?

**Comment by glaubitz** [comment]
Ah, right. I was already wondering why ```git submodule``` didn't list ```liblibc```. That explains it :).

**Comment by glaubitz** [comment]
@alexcrichton Oh, before you release. I think @karcherm has two fixes for sparc64 in ```libc``` ready.

**Comment by karcherm** [comment]
One of the patches @glaubitz is talking about is [https://github.com/rust-lang/libc/pull/1187]. I'm unsure whether it should be merged or not. The second one is work in progress to enable reliable stack overflow detection on sparc64. It is not ready yet, but possibly in a couple of hours.

**Comment by glaubitz** [comment]
@alexcrichton I would suggest to do a new ```libc``` now, so we can get this PR merged.

We can do another release once @karcherm has pushed his fixes.

**Comment by glaubitz** [comment]
I'm trying to update the ```libc``` crate now for this PR but that fails with:

```
glaubitz@epyc:..rust/rust> cargo update -p libc
    Updating crates.io index
error: no matching package named `core` found                                                                                                                                                
location searched: registry `https://github.com/rust-lang/crates.io-index`
required by package `backtrace-sys v0.1.27`
    ... which is depended on by `backtrace v0.3.11`
    ... which is depended on by `error-chain v0.12.0`
    ... which is depended on by `cargo_metadata v0.6.2`
    ... which is depended on by `miri v0.1.0 (/local_scratch/glaubitz/rust/rust/src/tools/miri)`
glaubitz@epyc:..rust/rust>
```

I made a fresh checkout to make sure my repo isn't corrupted but that didn't help.

Anyone got an idea what's wrong?

**Comment by alexcrichton** [comment]
Hm I think you may need to make sure to use the nightly cargo to update libc, but you can also just update one of the version requirements in this repository and then `./x.py build`

**Comment by glaubitz** [comment]
@alexcrichton @nagisa r?

File: src/librustc_data_structures/flock.rs [hunk 1]
```diff
@@ -16,6 +16,12 @@ cfg_if! {
         use std::os::unix::prelude::*;
         use libc;
 
+        pub const F_RDLCK: libc::c_short = libc::F_RDLCK as libc::c_short;
```
Comment (nagisa) [inline src/librustc_data_structures/flock.rs]: Why is it necessary to have these as constants here?

Comment (glaubitz) [reply src/librustc_data_structures/flock.rs]: Good point. I'll push an updated version in a second.

Comment (glaubitz) [reply src/librustc_data_structures/flock.rs]: Fixed :).

**Pull request merged**